frontend/include/chpl/parsing/parsing-queries.h

```diff
@@ -340,6 +339,47 @@ void setConfigSettings(Context* context, ConfigSettingsList keys);
 const
 ConfigSettingsList& configSettings(Context* context);
 
+/**
+  Given an ID, returns the attributes associated with the ID. Only
+  declarations can have associated attributes.
+ */
+const uast::Attributes* idToAttributes(Context* context, ID id);
+
+/**
+  Given an ID representing a mention of a symbol, and an ID representing
+  the symbol, determine if a deprecation warning should be produced
+  for the symbol. If so, the warning will be reported to the context
+  and returned. Returns nullptr if no deprecation warning is needed.
+
+  If a warning is produced, it will only be reported to the context
+  once per revision. The warning may not be reported if the context
+  is configured to suppress deprecation warnings, but it will still
+  be returned.
+
+  The mention ID may refer to any AST but will most often be an
+  Identifier. The target ID should refer to a NamedDecl. If it does
+  not, then nullptr is returned.
+*/
+const ErrorBase* deprecationWarningForId(Context* context, ID idMention,
```

Comment:
I just imagined that maybe a tool might like to see if an error was produced even in the event that it was not logged to the Context on that particular revision. Felt a bit better to me than just returning `void` and calling a query purely for the side effect.

For example, I was going to use the query + ids and mentions when writing some unit tests.

Reply:
For the record a few of the functions in `Context` for reporting errors currently also return the `ErrorBase*` for other logging / processing (it makes it easier to support patterns like `const auto reportedError = CHPL_REPORT(...)`. So to me, returning the (logged) error is not _that_ egregious, even though reporting macros aren't used here.

Reply:
Can you show me -- which functions & where is `reportedError = CHPL_REPORT` useful?

Reply:
So does this change our approach at all or do we still want to do `void reportDeprecationWarningForId` 😄.

Reply:
I'd still prefer `void reportDeprecationWarningForId`. I think we should only return the value in addition when we have a use case for it (rather than proactively) because I think it makes the function signature more prone to misinterpretation.